### ISSUE TITLE
OSS Index rate limit exceeded, disabling the analyzer

### ISSUE THREAD

**Issue by Muskan-0618** [dependency-check/DependencyCheck#6587]
We are experiencing OSS rate limit error in some of our scans. Complete error message is:- 

```
OSS Index rate limit exceeded, disabling the analyzer
exception: org.owasp.dependencycheck.analyzer.exception.AnalysisException: OSS Index rate limit exceeded, disabling the analyzer
org.owasp.dependencycheck.analyzer.OssIndexAnalyzer.analyzeDependency(OssIndexAnalyzer.java:151)
org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze(AbstractAnalyzer.java:131)
org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:88)
org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:37)
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
java.base/java.lang.Thread.run(Thread.java:834)
```
We are using the OSS Index credentials to trigger the scans but still getting this error if multiple scans(more than 4) are getting triggered at an interval of 1-2 mins.

Any form assistance would be greatly appreciated.

**Comment by aikebah** [comment]
ensure to persist and share the data folder of dependencycheck among those invocations to benefit from reusing recent (cached) responses from other scans on identical libraries

**Comment by jeremylong** [comment]
@aikebah I think we can do better on the caching for the OSS Index... but saving the data folder between scans is something far too few users do and it causes them issues.

**Comment by GunoH** [comment]
So is the OSS Index cached in the data folder?

The _disableOssIndexCache_ option listed [here](https://jeremylong.github.io/DependencyCheck/dependency-check-cli/arguments.html) suggests that there is some form of caching. That would be helpful, as I'm running in to the error of this issue quite frequently, while scanning several hundred projects in succession (using the [Single node database updater with multiple node “readers”](https://jeremylong.github.io/DependencyCheck/data/cacheh2.html) approach).

**Comment by aikebah** [comment]
@GunoH yes caches go underneath the datafolder. They are being cached by the OSSIndex API client (in folders underneath "oss_cache" inside the datafolder.

OSSIndex API client is AFAICT is caching the individual library results, so if your projects don't share that many libraries/version you might still run into issues with their ratelimiting, but if there is a large overlap an you share the full datafolder you would be expected to significantly lower the amount of calls to their API as successive scan would find most libs in cache and need to call the API only for unique to the new project libs and potentially (don't know how those are handled by their caching, whether they store/cache a not-found or not) the 'library coordinates not known in their database'.

The cache is configured by us to expire after 24hrs (assuming the cache folder exists or can be created)

https://github.com/dependency-check/DependencyCheck/blob/e462a236841771b58a4831991334340d8defbb6b/core/src/main/java/org/owasp/dependencycheck/data/ossindex/OssindexClientFactory.java#L92-L101

**Comment by GunoH** [comment]
Thanks.

Well, the projects do share many of the libraries, but they each use a fresh copy of the data directory that was saved in the single `--updateOnly` step, i.e., anything added to the data folder by the scans themselves (`--noupdate`) will not be there for the next project. I guess that explains why the caching of OSSIndex results is not that effective for me.

I suppose I'll have to rethink the strategy in the build pipeline, or otherwise disable OSSIndex.
